### ISSUE TITLE
Isobuild watch and build unnecessary files - under Meteor > v1.2

### ISSUE THREAD

**Issue by theosp** [meteor/meteor#7842]
[Issue posted first on Meteor forum.](https://forums.meteor.com/t/issues-with-isobuild-following-project-migration-from-v1-2-1-to-v1-3-1-4/29798)

I'd like to raise your attention to two key differences in the way isobuild works between v1.2.1 to v1.3.5.1:
1. Meteor watches packages files that are not referenced by package.js .
2. Meteor build plugins build files not referenced by package.js .
   These two differences caused me some issues when upgrading to v1.3.5.1 (before upgrading to v1.4, where the issue exist as well).

[I refined an example project that demonstrate the issues.](https://github.com/theosp/meteor-v1.2.1-to-v1.3.5.1-upgrade-issue)

In the repository, you can find two Meteor projects, one of v1.2.1 and another of v1.3.5.1 (also v1.4.1.1, where the results are similar to v1.3.5.1).

In each of these projects I've installed a package 'example:aaa-compiler' with build plugin that uses the legacy build plugins api (which my legacy code and 3rd party packages that I am using still use) to build files that ends with .aaa , [the build process adds to the bundle the js code: "console.log('I should be printed only once!')"](https://github.com/theosp/meteor-v1.2.1-to-v1.3.5.1-upgrade-issue/blob/master/aaa-compiler/build-plugin.js#L9).

[example:aaa-compiler's package.js](https://github.com/theosp/meteor-v1.2.1-to-v1.3.5.1-upgrade-issue/blob/master/aaa-compiler/package.js) has only the build plugin, and reference only to the build plugin file.

In the package folder there's a /docs/example-aaa folder [with example aaa file](https://github.com/theosp/meteor-v1.2.1-to-v1.3.5.1-upgrade-issue/tree/master/aaa-compiler/docs/example-aaa).

That example file is symlinked to the app root in both projects.
# Running results:
## V1.2.1 - Web

![1 2 1-web](https://cloud.githubusercontent.com/assets/207494/18938559/032323a8-862c-11e6-946a-72f55e1bd182.png)
## V1.2.1 - Server

![1 2 1-server](https://cloud.githubusercontent.com/assets/207494/18938562/10a2416c-862c-11e6-8ac6-0584f15f6229.png)
## V1.3.5.1 - Web

![1 3 5 1-web](https://cloud.githubusercontent.com/assets/207494/18938575/1efcb3a0-862c-11e6-9bad-3136d342316c.png)
## V1.3.5.1 - Server

![1 3 5 1-server](https://cloud.githubusercontent.com/assets/207494/18938582/252822fa-862c-11e6-926f-d601e4a2c95e.png)
# Workaround:

As a workaround, for now, [I added .# prefix to the folder that shouldn't be watched](https://github.com/meteor/meteor/blob/devel/tools/isobuild/bundler.js#L176). This is a very hacky/not-documented workaround, which make the folders hidden for regular operations which might interfere with the project maintenance.
# Questions:
1. Why packages files that are out of package.js scope are watched? Two issues with that:

a. If I have big documentation folder (think something > 5mb with lots of files) why should it be watched?

b. Maybe this is a cause for increase in build time in > v1.2?

c. I used to have common code residing in packages which I symlinked to the main app, now I can't do that without getting warnings from the builder (since it now watch duplicates).
1. Is there a reason why these out-of-scope files are built? Test folders and documentation folders under a package folder can't assume they are isolated and out of scope any longer?
2. Is this a bug or intended behaviour? I couldn't find any serious discussion about this issue, or any place where it's documented.

Thanks,
-Daniel

**Comment by lucfranken** [comment]
Ah see you posted it here as well next to the forums. Try: https://github.com/meteor/meteor/pull/7668

**Comment by theosp** [comment]
A test added in the repository to 1.4.2-beta.7 , same issue.

**Comment by theosp** [comment]
A test added in the repository to 1.4.2-rc3 , same issue.

**Comment by theosp** [comment]
@mitar is it possible that this issue relates to your need for .* files ignoring mentioned here https://github.com/meteor/meteor/pull/7668#issuecomment-255978149 or is this another usecase?

**Comment by abernix** [comment]
From an issue triage standpoint, I haven't investigated this beyond confirming this reproduction however, it's worth pointing out in your reproduction that when you first run it, it actually only prints the message once (reported as `app.js`):

    I should be printed only once!

If you then stop the server and start it **again**, only then will you start seeing the second message from `example_aaa-compiler.js`.  While I'm fairly sure the duplication is due to the differences you noted, I found this to be confusing in running your reproduction.

**Comment by hwillson** [comment]
There have been several file watcher changes made since Meteor 1.4.2. Please try things out with a current version of Meteor (>= 1.6), and post back if you're still having this problem. Closing for now - thanks!
